Make GPU manager skip records when nothing scheduled on input stream

therault opened the pull request:
Currently, the GPU manager systematically records 3 polling events (or more for TTG) per task: one after kernel_push, one after kernels submission and one after kernel_pop. The event record after the kernel_push is needed if any input data is not yet on the GPU, even if that task did not schedule an actual cudaMemcpyAsync, because the cudaMemcpyAsync it depends on is already executing, and we use the serialization of the input stream to guarantee order.

However, in many cases (typically O(nt^3-nt^2) tasks for a GEMM that fits on GPU memory, or for a POTRF that fits in GPU memory), *all* data are already on the GPU. In that case, recording an event and waiting for it to exit the queue can be unnecessary overhead.

This PR skips event recording after the input stream if *all* input data is already on the GPU.

Pro:
  - reduces number of events to poll
Con:
  - will increase the cases of out-of-order execution.

@QingleiCao has tested the PR with variable tile size, and some benefit is observed for small tiles. No effect (detrimental or beneficial) is observed for large tiles on DPLASMA. @devreal wanted to test with TTG, where this is expected to bring more benefit (very small work in many tasks) and this PR should simplify him testing the combination of multiple branches.

therault committed:
Improve efficiency of stream management in GPU.

When a GPU kernel had all its data already on the GPU, we would still schedule
a record event on the input stream. That would delay the scheduling of the kernel
after all asyncMemcpy that had been scheduled before complete their execution,
reducing parallelism between execution and I/O.

With this change, the aim is to entirely skip recording the event on the input
stream and directly schedule the ready-to-execute kernel on an exec stream of
the GPU.

Note that the behavior of data already in transfer is unchanged: no additional
transfer is scheduled, but the task needs to wait on the input stream progress
to be ready to execute.

bosilca reviewed (commented):
By this logic all submission functions should now return a positive number, including the DTD GPU submission, the coroutines and the all the user-provided stage_in and out functions.

devreal commented on parsec/mca/device/device_gpu.c:
```diff
@@ -38,6 +38,13 @@ static int parsec_gpu_profiling_initiated = 0;
 int parsec_gpu_output_stream = -1;
 int parsec_gpu_verbosity;
 
+/* The return value of these functions is either a parsec_hook_return_t for <= 0 values,
```
The semantics are different from the existing function: `parsec_advance_task_function_t` returns a `parsec_hook_return_t` while the new function returns either `parsec_hook_return_t` or a positive number if something was scheduled on the stream. We opted to make that a new typedef.

devreal commented:
@therault pretty please :)

therault commented on parsec/mca/device/device_gpu.c:
```diff
@@ -1882,6 +1890,13 @@ parsec_device_progress_stream( parsec_device_gpu_module_t* gpu_device,
 
   schedule_task:
     rc = progress_fct( gpu_device, task, stream );
+    if(0 == rc) {
```
I think it's dubious that the callback function started some profiling event, yet did nothing (which is what 0 == rc means). But to be on the safe side, I have included that code. I expect `task->prof_key_end` to be -1 all the time, though.

bosilca replied:
I was not referring to the internal profiling logic, this is for GPU there is no such thing we are based on events. But if the user is trying to understand the progress of a task in the GPU it will likely count the events to see the status of the task, and by not generating an event on the stage_in this will confuse the counting.

bosilca commented on parsec/mca/device/device_gpu.c:
```diff
@@ -1949,6 +1957,20 @@ parsec_device_progress_stream( parsec_device_gpu_module_t* gpu_device,
+            rc = task->complete_stage(gpu_device, &task, stream);
+        *out_task = task;
+        return rc;
+    }
```
How about we rely on the already existing code for returning a completed task from the stream that is a little above in this function. Instead of replicating the code here, and creating two identical return points (both returning the task can be advanced to the next stage), we merge these two into a single. For this we can replace the code above with
```c
task->last_status = PARSEC_HOOK_RETURN_DONE;
goto withdraw_the_task;
```
and then add the `withdraw_the_task` label at line 1910 in my version of the file, basically right after `stream->end = (stream->end + 1) % stream->max_events;`.

devreal replied:
But the code below 1910 does modify `stream->tasks`, which we should not touch here. Plus, it won't help readability.

devreal commented:
Tested this on a single A100 node and saw no problems. Please provide another review so this can be merged.

bosilca reviewed (commented):
I am not certain I understand how this works. `parsec_device_progress_stream` is the core for all GPU interactions, used not only for push (which is what this PR tries to optimize) but also for exec and pop. Most of these interactions return `PARSEC_HOOK_RETURN_DONE`, aka 0, when they successfully complete their work. With this PR the return from the `progress_fct` is used to define if the completion event exists or not, and because most of the submission functions already return 0 (because `DONE`) their completion event will not be generated and instead call the next step automatically. The problem here is that the kernels are enqueued and they will be waited on eventually (which translates into a lack of impact on the performance) , but we are losing the sequential ordering. In other words this seem to work only because successive kernels on the same output tile end up being submitted far from each other , the result will therefore look correct most of the time, but will not really be valid.

devreal commented:
> With this PR the return from the progress_fct is used to define if the completion event exists or not, and because most of the submission functions already return 0 (because DONE) their completion event will not be generated and instead call the next step automatically. 

No. `parsec_device_kernel_exec` returns `1` if a kernel was submitted and thus we insert an event. Zero is only returned for push and pop if there is nothing to push or pop (or we are depending on an ongoing transfer), allowing to go to the next step without waiting for other (unrelated) transfers to complete. 

See https://github.com/ICLDisco/parsec/pull/681/changes#diff-31674b6b990f60627dec43ecfc8bfd96bbd1f6be0535a4008da26b8f5ee87becR2180

bosilca commented:
You are correct, I missed that specific return. I understand why some might see an appeal in this optimization despite the fact that the benefit seems minimal, but I have two concerns:
1. don't like the fact that we are creating yet another opportunity for losing tasks ordering, one based not on task's priority (that will trigger the sorting of tasks on each step of the stream by its priority) but on data availability.
2. The code is messy. We allow the user to attach a function that is of type `parsec_advance_task_function_t` (which return no information about submission of a kernel) and then we wrap this around a function with a barely different signature `parsec_gpu_step_function_t` where we forcefully ignore the 0 return from the user function and replace it with a 1 to indicate a kernel has been submitted (which we can only assume of being true).

How about:
1. we drop the `parsec_gpu_step_function_t` typedef and instead change the meaning of the `parsec_advance_task_function_t` functions to return any positive value to indicate work has been enqueued on the stream; and
2. we add an MCA parameter to allow/forbid this optimization
3. I would also change the name of the `parsec_advance_task_function_t` to indicate it only applies to devices, such as `parsec_device_submit_task_function_t`

devreal commented:
> don't like the fact that we are creating yet another opportunity for losing tasks ordering, one based not on task's priority (that will trigger the sorting of tasks on each step of the stream by its priority) but on data availability.

You mean that we schedule tasks before all prior transfers are completed? Because once it hits the execution streams we're going out-of-order anyway. If the data copy accounting is correct (i.e., under-transfer and provenance are correctly handled) this should not be an issue. IMO, it is more important to keep the device busy than to strictly enforce task priority ordering, esp if we have many tasks with all data available and a few that wait for transfers.

> How about:
>
> 1. we drop the parsec_gpu_step_function_t typedef and instead change the meaning of the parsec_advance_task_function_t functions to return any positive value to indicate work has been enqueued on the stream; and

`parsec_advance_task_function_t` is part of the public API today. I would not want to change the public API in this PR. That would require us to change DPLASMA and TTG and makes things messy. My understanding is that today the submit callback return either `PARSEC_HOOK_RETURN_DONE` or `PARSEC_HOOK_RETURN_AGAIN` if a kernel was submitted. Neither of them are non-zero positive integers so this would be a break.

We could change the signature of `parsec_gpu_step_function_t` to differentiate it from the existing `parsec_advance_task_function_t`. The return value could be `int64_t` instead of `int`. That would make it easier to catch cases where we use a `parsec_advance_task_function_t` for a `parsec_gpu_step_function_t`. 

> 2. we add an MCA parameter to allow/forbid this optimization

We can add an option to force serialization on the input stream, at the cost of code complexity.

> 3. I would also change the name of the parsec_advance_task_function_t to indicate it only applies to devices, such as parsec_device_submit_task_function_t

That is a different change and changes the public API (`parsec_advance_task_function_t` existed before this change). I'd rather not do that at this point.

devreal committed:
Add MCA parameter device_skip_empty_events

By default, we are skipping events if there is nothing to wait for.
By setting device_skip_empty_events=0 we can restore the old behavior
of always recording an event on input and output streams.

Signed-off-by: Joseph Schuchart <joseph.schuchart@stonybrook.edu>

devreal commented:
@bosilca As discussed in the last meeting, I added an MCA parameter `device_skip_empty_events` to force events to always be recorded on the input and output streams even if they are not strictly needed, effectively disabling the effects of this PR.

bosilca commented:
Once they hit the execution streams they go out-of-order only because we have multiple concurrent streams, but the task priority still drive their dispatch onto the streams. I know the priority was just a hint, now it becomes even less important. Anyway, I don't have much concerns about the capability I have concerns about the implementation.

It does not change the public API, that part was never really supposed to be exposed to the users. But if we keep adding variations of the same APIs with minimal differences we are just making things more complicated for the parsec developers. I think that as we prepare for a major release, now it is the time to make changes to the APIs that the users can access.

devreal commented:
> It does not change the public API, that part was never really supposed to be exposed to the users.

The `parsec_advance_task_function_t` callbacks are the public API. They are provided by the application. I really don't see the problem of having two different callback types since they are not mixed: `parsec_gpu_step_function_t` is internal while `parsec_advance_task_function_t` is public. One calls the other, but that's about it. Having the public callback strictly return `parsec_hook_return_t` makes sense because it's a well-defined value range. If we want the application to signal that nothing was submitted we can extend `parsec_hook_return_t` to encode that without breaking existing code.

Making a larger change that involves fixing DPLASMA and who knows what other software in preparation of the release in this PR is feature creep. I'm afraid I don't have the cycles for it and will have to stay on our PaRSEC fork for TTG for the time being.

bosilca committed:
device/gpu: preserve hook status across skipped events

When parking a GPU task behind an event, keep the progress return code in
last_status so deferred AGAIN paths retry the same stage at event completion
instead of advancing the task.

Also propagate negative submit hook returns from the execution stage
immediately, while keeping successful submissions on the event-record path.
Clarify that device_skip_empty_events only applies to input/output stream
events, not successful execution submissions.

Signed-off-by: George Bosilca <gbosilca@nvidia.com>

bosilca committed:
Protect GPU D2D source ownership

Avoid moving an OWNED GPU copy onto the clean GPU LRU when the last D2D
reader is released. D2D readers do not transfer ownership, so dirty
GPU-only data must remain visible to the W2R backup path before it can be
reused.

Also make the D2D fast source path reject invalid or under-transfer
copies before acquiring a reader. When a potential GPU source exists but
is not ready, only fall back to the CPU copy if the host copy has the
expected version and is ready/resident; otherwise return
PARSEC_HOOK_RETURN_AGAIN before changing ownership state.

Signed-off-by: George Bosilca <gbosilca@nvidia.com>

devreal commented:
Thanks @bosilca @therault!
